### ISSUE TITLE
stop requiring event arguments to extend EventArgs in event raisers for .NET Standard

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#361]
From @bondsbw in #30:

> I know I'm coming in late to this discussion, but something to consider (I apologize if I missed this in past discussion):\
> .NET 4.5 no longer constrains EventHandler to using something that derives from EventArgs. TEventArgs can now be anything.
> 
> Of course compiling for .NET 4.5 may not make sense depending on the goals of the project.
> But I made a few alterations to the source to bump it to 4.5 and remove the constraint,
> and was able to get everything working as expected (other than Silverlight projects which I did not load).

**Comment by 9swampy** [comment]
Thank-you. It's taken a long time to make sense of using #30 but thanks to your last comment I got it to work with 2.0.0beta009 in a 4.5 project.

For:
public delegate void NewFrameEventHandler(object sender, NewFrameEventArgs eventArgs);
Exposed via:
event NewFrameEventHandler NewFrame;

Either of the following worked:
.NewFrame += Raise.With<NewFrameEventHandler>(new object[] { this, new NewFrameEventArgs(bitmap) });
.NewFrame += Raise.With<NewFrameEventHandler>(this, new NewFrameEventArgs(bitmap));

[Slightly off topic as in my case NewFrameEventArgs does still inherit from EventArgs]

**Comment by blairconrad** [comment]
@9swampy, I'm glad you were able to raise the events. Thanks for the feedback.

**Comment by adamralph** [comment]
So do we actually have to do anything? Or will 2.0 just do this already?

**Comment by blairconrad** [comment]
I'll give it a whirl soonish and get back to you, @adamralph.

**Comment by blairconrad** [comment]
@adamralph, the 2.0 betas do not do what we want.

If there were a delegate defined, and the event defined in terms of the delegate, we could raise the event, just as with any delegate-defined event, by using this syntax:

```c#
public delegate void DoesNotExtendEventHandler(object sender, DoesNotExtendEventArgs args);

public interface IRaiseEvents
{
    event DoesNotExtendEventHandler AnEvent;
}

…
fake.AnEvent += Raise.With<DoesNotExtendEventHandler>(this, new DoesNotExtendEventArgs());
```

If we omit the `<DoesNotExtendEventHandler>`, nothing compiles because `DoesNotExtendEventArgs` isn't convertible to `EventArgs`.

Removing the restriction on `Raise` that the event args extend `EventArgs` is not possible because in 4.0, generic `EventHandler`s only work for event argument types that extend `EventArgs`.
Targeting 4.5 should fix that problem.
It's possible that there's some other trickery that could be performed to avoid having to target 4.5, but I've not gone looking too hard. When I removed the requirement for the `.Now`, I did look for a way to avoid having to specify the delegate type as in the code above, and I was not successful. Of course, I was looking for a universally-applicable solution, not something that would apply specifically to an event with signature
`(object sender, TEventArgs eventArgs)` (with no restriction on `TEventArgs`).

**Comment by adamralph** [comment]
Is it worth creating a separate net45 assembly in the nupkg for this?

**Comment by blairconrad** [comment]
It seems like a lot of effort for the benefit. Off the top of my head, I'm not sure whether there would be anything else we could leverage.

Then again, it would provide a nice experience for users who are using what appears to be the new blessed event format. I'm not against it, but amn't in a huge rush…

If we decide to do so, though, should we examine mechanisms for building these without the manual "adding files to the project" like we do for 3.5? Shared projects seem to be the wave of the present for this kind of thing, but that would require all contributors (and the CodeBetter CI) to use VS 2015 or 2013 with an extension, no?

**Comment by adamralph** [comment]
We can also use this method https://github.com/xbehave/xbehave.net/blob/7c49a81af810ee7a9902ffea6008ed7a046d1f2a/src/Xbehave.Sdk.Net45/Xbehave.Sdk.Net45.csproj#L71-L75

Shared projects would seem to be the obvious answer, but I know at least one project (xUnit.net) which tried them but ultimately abandoned them due to difficulties.

**Comment by thomaslevesque** [comment]
Shared projects are the way to go IMO. I've been using them for a while on projects at work, and haven't regretted it. The only drawback is that some tools (e.g. ReSharper) don't always handle them properly, but it got better in the recent versions.

> that would require all contributors (and the CodeBetter CI) to use VS 2015 or 2013 with an extension, no?

Is this really an issue? Contributors are usually enthusiasts that don't wait too long before adopting a new VS release...

**Comment by blairconrad** [comment]
@thomaslevesque, I don't think requiring VS2015 is a deal-breaker. I just thought we might want to be aware of it when making the switch.
(And if we make the switch, we could start using bling strings and other new C# goodness…)

**Comment by thomaslevesque** [comment]
> (And if we make the switch, we could start using bling strings and other new C# goodness…)

Sounds good to me :smile:

**Comment by adamralph** [comment]
One thing to bear in mind before adopting C# 6, is that StyleCop.MSBuild will have to be removed since it doesn't (and almost certainly never will) support C# 6. A potential replacement is https://github.com/DotNetAnalyzers/StyleCopAnalyzers.

**Comment by thomaslevesque** [comment]
> One thing to bear in mind before adopting C# 6, is that StyleCop.MSBuild will have to be removed since it doesn't (and almost certainly never will) support C# 6. A potential replacement is https://github.com/DotNetAnalyzers/StyleCopAnalyzers.

Well, to be honest, I'm not a big fan of StyleCop (too finicky for my taste), so I wouldn't be very upset if we dropped it completely :wink:. But if you want to keep it, then yes, StyleCopAnalyzers seems to be a good replacement; looks like all the rules are implemented, most of them with a corresponding codefix (which is certainly an improvement over StyleCop)

**Comment by adamralph** [comment]
It's only as finicky as the rules you switch on, and that can be a team decision

E.g. in FIE, we have almost all switched on. In scriptcs, we only have a handful.

I am always surprised how people love to blame the tool. :smile:

**Comment by thomaslevesque** [comment]
> It's only as finicky as the rules you switch on, and that can be a team decision

True

**Comment by blairconrad** [comment]
More rules! All the rules!

**Comment by blairconrad** [comment]
In all seriousness, I don't agree with all the rules (it still feels weird to me to specify `this.` all the time), but I have really come to enjoy that the rules are there. I like the consistent spacing and line endings and ordering of members and whatnot. Ah well, we've veered off course a little.

**Comment by blairconrad** [comment]
There didn't seem to be much appetite for making a separate .NET 4.5 package just for this, but we _do_ have a .NET Standard 1.6 package that could support non-EventArg-extending type params for EventHandler…

**Comment by blairconrad** [comment]
Marking **in-progress** until someone says we don't want to do ti.

**Comment by adamralph** [comment]
When we switch to the new csproj, which can multi-target, we should switch on a .net45 output which also has this.

**Comment by thomaslevesque** [comment]
> When we switch to the new csproj, which can multi-target, we should switch on a .net45 output which also has this.

My thoughts exactly

**Comment by blairconrad** [comment]
If only you'd let me fix #991, we could have the .net45 project now! :wink:

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 3.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.1.0).
